Mathlib/Data/Set/Lattice.lean

```diff
@@ -0,0 +1,2522 @@
+/- ./././Mathport/Syntax/Translate/Expr.lean:107:6: warning: expanding binder group (i j) -/
+theorem unionᵢ₂_congr {s t : ∀ i, κ i → Set α} (h : ∀ i j, s i j = t i j) :
+    (⋃ (i) (j), s i j) = ⋃ (i) (j), t i j :=
+  unionᵢ_congr fun i => unionᵢ_congr <| h i
+#align set.Union₂_congr Set.unionᵢ₂_congr
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:107:6: warning: expanding binder group (i j) -/
+/- ./././Mathport/Syntax/Translate/Expr.lean:107:6: warning: expanding binder group (i j) -/
+theorem interᵢ₂_congr {s t : ∀ i, κ i → Set α} (h : ∀ i j, s i j = t i j) :
+    (⋂ (i) (j), s i j) = ⋂ (i) (j), t i j :=
+  interᵢ_congr fun i => interᵢ_congr <| h i
+#align set.Inter₂_congr Set.interᵢ₂_congr
+
+theorem bunionᵢ_eq_unionᵢ (s : Set α) (t : ∀ x ∈ s, Set β) :
+    (⋃ x ∈ s, t x ‹_›) = ⋃ x : s, t x x.2 :=
+  supᵢ_subtype'
+#align set.bUnion_eq_Union Set.bunionᵢ_eq_unionᵢ
+
+theorem binterᵢ_eq_interᵢ (s : Set α) (t : ∀ x ∈ s, Set β) :
+    (⋂ x ∈ s, t x ‹_›) = ⋂ x : s, t x x.2 :=
+  infᵢ_subtype'
+#align set.bInter_eq_Inter Set.binterᵢ_eq_interᵢ
+
+theorem unionᵢ_subtype (p : α → Prop) (s : { x // p x } → Set β) :
+    (⋃ x : { x // p x }, s x) = ⋃ (x) (hx : p x), s ⟨x, hx⟩ :=
+  supᵢ_subtype
+#align set.Union_subtype Set.unionᵢ_subtype
+
+theorem interᵢ_subtype (p : α → Prop) (s : { x // p x } → Set β) :
+    (⋂ x : { x // p x }, s x) = ⋂ (x) (hx : p x), s ⟨x, hx⟩ :=
+  infᵢ_subtype
+#align set.Inter_subtype Set.interᵢ_subtype
+
+theorem binterᵢ_empty (u : α → Set β) : (⋂ x ∈ (∅ : Set α), u x) = univ :=
+  infᵢ_emptyset
+#align set.bInter_empty Set.binterᵢ_empty
+
+theorem binterᵢ_univ (u : α → Set β) : (⋂ x ∈ @univ α, u x) = ⋂ x, u x :=
+  infᵢ_univ
+#align set.bInter_univ Set.binterᵢ_univ
+
+@[simp]
+theorem bunionᵢ_self (s : Set α) : (⋃ x ∈ s, s) = s :=
+  Subset.antisymm (unionᵢ₂_subset fun _ _ => Subset.refl s) fun _ hx => mem_bunionᵢ hx hx
+#align set.bUnion_self Set.bunionᵢ_self
+
+@[simp]
+theorem unionᵢ_nonempty_self (s : Set α) : (⋃ _h : s.Nonempty, s) = s := by
+  rw [unionᵢ_nonempty_index, bunionᵢ_self]
+#align set.Union_nonempty_self Set.unionᵢ_nonempty_self
+
+-- TODO(Jeremy): here is an artifact of the encoding of bounded intersection:
+-- without dsimp, the next theorem fails to type check, because there is a lambda
+-- in a type that needs to be contracted. Using simp [eq_of_mem_singleton xa] also works.
+theorem binterᵢ_singleton (a : α) (s : α → Set β) : (⋂ x ∈ ({a} : Set α), s x) = s a :=
+  infᵢ_singleton
+#align set.bInter_singleton Set.binterᵢ_singleton
+
+theorem binterᵢ_union (s t : Set α) (u : α → Set β) :
+    (⋂ x ∈ s ∪ t, u x) = (⋂ x ∈ s, u x) ∩ ⋂ x ∈ t, u x :=
+  infᵢ_union
+#align set.bInter_union Set.binterᵢ_union
+
+theorem binterᵢ_insert (a : α) (s : Set α) (t : α → Set β) :
+    (⋂ x ∈ insert a s, t x) = t a ∩ ⋂ x ∈ s, t x := by simp
+#align set.bInter_insert Set.binterᵢ_insert
+
+-- TODO(Jeremy): another example of where an annotation is needed
```

Comment:
No annotation is here. The proof used to be `by rw insert_of_has_insert; simp` five years ago.